### PR TITLE
KNN unique implementation

### PR DESCRIPTION
**Pull request by mayurdb**
A priority queue based implementation of the KNN which return k unique elements.
Steps - 
 1. Find k unique nearest neighbors in each of the partition.
 2. Merge the lists from all partitions into a single list of size k based on the ordering.

### CODE REVIEW

File: src/main/scala/com/spark3d/utils/BoundedUniquePriorityQueue.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  override def size: Int = underlying.size
+
+  override def ++=(xs: TraversableOnce[A]): this.type = {
+    xs.foreach { this += _ }
+    this
+  }
+
+  override def +=(elem: A): this.type = {
+    if (!underlying.map(x => x.center).toList.contains(elem.center)) {
```
[the remark this answers is not in the document]
Comment (JulienPeloton) [reply src/main/scala/com/spark3d/utils/BoundedUniquePriorityQueue.scala]: Hum not sure I understand the logic here. What would you propose instead?

**Comment by JulienPeloton** [comment]
There is one conflict also to resolve before the merge.

**Comment by mayurdb** [comment]
Fixed the merge conflict

**Comment by JulienPeloton** [comment]
OK sounds good to me. I'm OK for the merge if you agree.

**Comment by mayurdb** [comment]
Yes, we can merge this!

**Pull request merged**